tcf_website/views/browse.py

```diff
@@ -75,36 +75,26 @@ def course_view(request, course_id):
 
     course = Course.objects.get(pk=course_id)
     latest_semester = Semester.latest()
-
-    # Get instructors that have taught the course in the most recent
-    # semester.
-    recent_instructor_pks = course.section_set.filter(
-        semester=latest_semester).values_list(
-        'instructors', flat=True).distinct()
-    recent_instructors = Instructor.objects.filter(
-        pk__in=recent_instructor_pks)
-    # Add ratings and difficulties
-    for instr in recent_instructors:
-        instr.rating = instr.average_rating_for_course(course)
-        instr.difficulty = instr.average_difficulty_for_course(course)
-        instr.gpa = CourseInstructorGrade.objects.filter(
-            course=course, instructor=instr).aggregate(
-            Avg('average')).get('average__avg')
-
-    # Get instructors that haven't taught the course this semester.
-    old_instructor_pks = course.section_set.exclude(
-        semester=latest_semester).exclude(
-        instructors__pk__in=recent_instructor_pks).values_list(
-        'instructors',
-        flat=True).distinct()
-    old_instructors = Instructor.objects.filter(pk__in=old_instructor_pks)
-    # Add ratings and difficulties
-    for instr in old_instructors:
-        instr.rating = instr.average_rating_for_course(course)
-        instr.difficulty = instr.average_difficulty_for_course(course)
-        instr.gpa = CourseInstructorGrade.objects.filter(
-            course=course, instructor=instr).aggregate(
-            Avg('average'))['average__avg']
+    instructors = Instructor.objects\
+        .filter(section__course=course).distinct()\
+        .annotate(
+            gpa=Avg('courseinstructorgrade__average',
+                    filter=Q(courseinstructorgrade__course=course)),
+            difficulty=Avg('review__difficulty', filter=Q(review__course=course)),
+            rating=(
+                Avg('review__instructor_rating', filter=Q(review__course=course)) +
+                Avg('review__enjoyability', filter=Q(review__course=course)) +
+                Avg('review__recommendability', filter=Q(review__course=course))
+            ) / 3,
```

Comment:
We have an `average_rating_for_course(self,course)` method in the Instructor model that basically does this — it might be better for readability if we just used that instead. Is there a performance improvement by doing this instead?

Reply:
The method you're referring to works for an *instance* of `Instructor`, but not for a *queryset* of `Instructor`s. 
![image](https://user-images.githubusercontent.com/7676354/99767651-bc318880-2b46-11eb-9aaa-8d7b14b1dc05.png)
We need to work with querysets so that we don't access the database for every instance.